File: python/matx/vision/tv_transforms/crop.py
# Copyright 2022 ByteDance Ltd. and/or its affiliates.
#
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

from os import sync
from typing import Tuple, Union, Sequence, Dict, Any, List, overload
import numbers
import random
import sys
matx = sys.modules['matx']
from .. import ASYNC, BORDER_CONSTANT, BORDER_REPLICATE, BORDER_REFLECT_101, BORDER_REFLECT
from .. import CropOp, PadWithBorderOp
from ._base import BaseInterfaceClass, BatchBaseClass


class CenterCrop(BaseInterfaceClass):
    def __init__(self,
                 size: List[int],
                 device_id: int = -2,
                 sync: int = ASYNC) -> None:
        super().__init__(device_id=device_id, sync=sync)
        # check size
        if len(size) == 1:
            self._size: Tuple[int, int] = (size[0], size[0])
        elif len(size) == 2:
            self._size: Tuple[int, int] = (size[0], size[1])
        else:
            assert False, "Crop size value should be an integer or a list/tuple with length 2."

    def __call__(self, device: Any, device_str: str, sync: int) -> Any:
        return CenterCropImpl(device, device_str, self._size, sync)


class CenterCropImpl(BatchBaseClass):
    def __init__(self,
                 device: Any,
                 device_str: str,
                 size: Tuple[int, int],
                 sync: int = ASYNC) -> None:
        super().__init__()
        self.size: Tuple[int, int] = size
        self.device_str: str = device_str
        self.sync: int = sync
        fill = (0, 0, 0)
        padding_mode = BORDER_CONSTANT
        self.pad_op: PadWithBorderOp = PadWithBorderOp(device, fill, padding_mode)
        self.crop_op: CropOp = CropOp(device)
        self.name: str = "CenterCropImpl"

    def get_crop_params(self, h: int, w: int) -> Tuple[int, int, int, int]:
        th, tw = self.size
        x = (w - tw) // 2
        y = (h - th) // 2
        return y, x, th, tw

    def get_pad_params(self, h: int, w: int) -> Tuple[int, int, bool]:
        th, tw = self.size
        h_pad, w_pad = 0, 0
        if th > h:
            h_pad = int((1 + th - h) / 2)
        if tw > w:
            w_pad = int((1 + tw - w) / 2)
        need_pad = (h_pad + w_pad) > 0
        return h_pad, w_pad, need_pad

    def _process_crop_op(self, imgs: List[matx.NDArray]) -> List[matx.NDArray]:
        batch_size = len(imgs)
        need_pad = False
        h_pads, w_pads = [], []
        padded_height, padded_width = [], []
        for i in range(batch_size):
            h, w = imgs[i].shape()[:2]
            h_pad, w_pad, need_pad_tmp = self.get_pad_params(h, w)
            need_pad |= need_pad_tmp
            h_pads.append(h_pad)
            w_pads.append(w_pad)
            padded_height.append(h + 2 * h_pad)
            padded_width.append(w + 2 * w_pad)
        if need_pad:
            imgs = self.pad_op(imgs, h_pads, h_pads, w_pads, w_pads, sync=self.sync)
        crop_x, crop_y, crop_w, crop_h = [], [], [], []
        for i in range(batch_size):
            y, x, h, w = self.get_crop_params(padded_height[i], padded_width[i])
            crop_x.append(x)
            crop_y.append(y)
            crop_w.append(w)
            crop_h.append(h)
        imgs = self.crop_op(imgs, crop_x, crop_y, crop_w, crop_h, sync=self.sync)
        return imgs

    def _process(self, imgs: List[matx.NDArray]) -> List[matx.NDArray]:
        return self._process_crop_op(imgs)

    def __repr__(self) -> str:
        return self.name + '(size={0}, device={1}, sync={2})'.format(
            self.size, self.device_str, self.sync)


class RandomCrop(BaseInterfaceClass):
    def __init__(self,
                 size: List[int],
                 padding: List[int],
                 pad_if_needed: bool = False,
                 fill: List[int] = [0],
                 padding_mode: str = "constant",
                 device_id: int = -2,
                 sync: int = ASYNC) -> None:
        super().__init__(device_id=device_id, sync=sync)
        # check size
        if len(size) == 1:
            self._size: Tuple[int, int] = (size[0], size[0])
        elif len(size) == 2:
            self._size: Tuple[int, int] = (size[0], size[1])
        else:
            assert False, "Crop size value should be an integer or a list/tuple with length 2."
        # check padding
        if padding is None:
            self._padding: Tuple[int, int, int, int] = (0, 0, 0, 0)
        elif len(padding) == 1:
            self._padding: Tuple[int, int, int, int] = (
                padding[0], padding[0], padding[0], padding[0])
        elif len(padding) == 2:
            self._padding: Tuple[int, int, int, int] = (
                padding[0], padding[1], padding[0], padding[1])
        elif len(padding) == 4:
            self._padding: Tuple[int, int, int, int] = (
                padding[0], padding[1], padding[2], padding[3])
        else:
            assert False, "Padding must be None or a 1, 2 or 4 element tuple.."
        self._pad_if_needed: bool = pad_if_needed
        # check fill
        if len(fill) == 1:
            self._fill: Tuple[int, int, int] = (fill[0], fill[0], fill[0])
        elif len(fill) == 3:
            self._fill: Tuple[int, int, int] = (fill[0], fill[1], fill[2])
        else:
            assert False, "fill value should be a 1 or 3 element tuple."
        # check padding_mode
        assert padding_mode in ["constant", "edge", "reflect",
                                "symmetric"], "padding_mode should be constant, edge, reflect or symmetric."
        self._padding_mode: str = padding_mode

    def __call__(self, device: Any, device_str: str, sync: int) -> Any:
        return RandomCropImpl(
            device,
            device_str,
            self._size,
            self._padding,
            self._pad_if_needed,
            self._fill,
            self._padding_mode,
            sync)


class RandomCropImpl(BatchBaseClass):
    def __init__(self,
                 device: Any,
                 device_str: str,
                 size: Tuple[int, int],
                 padding: Tuple[int, int, int, int],
                 pad_if_needed: bool,
                 fill: Tuple[int, int, int],
                 padding_mode: str,
                 sync: int) -> None:
        super().__init__()
        self.device_str: str = device_str
        self.size: Tuple[int, int] = size
        self.padding: Tuple[int, int, int, int] = padding
        self.pad_if_needed: bool = pad_if_needed
        self.fill: Tuple[int, int, int] = fill
        self.padding_mode: str = padding_mode
        self.sync: int = sync
        self.crop_op: CropOp = CropOp(device)
        torch_padding_mode: Dict[str, str] = {
            "constant": BORDER_CONSTANT,
            "edge": BORDER_REPLICATE,
            "reflect": BORDER_REFLECT_101,
            "symmetric": BORDER_REFLECT
        }
        self.pad_op: PadWithBorderOp = PadWithBorderOp(
            device, self.fill, torch_padding_mode[self.padding_mode])
        self.name: str = "RandomCrop"

    def get_crop_params(self, h: int, w: int) -> Tuple[int, int, int, int]:
        th, tw = self.size
        if w == tw and h == th:
            return 0, 0, h, w
        i = 0
        j = 0
        if h - th > 0:
            i = random.randint(0, h - th)
        else:
            i = random.randint(h - th, 0)
        if w - tw > 0:
            j = random.randint(0, w - tw)
        else:
            j = random.randint(w - tw, 0)
        return i, j, th, tw

    def get_pad_params(self, h: int, w: int) -> Tuple[int, int, int, int, bool]:
        left_p, top_p, right_p, bot_p = self.padding
        if self.pad_if_needed and h + top_p + bot_p < self.size[0]:
            h_pad = self.size[0] - h - top_p - bot_p
            top_p += h_pad
            bot_p += h_pad
        if self.pad_if_needed and w + left_p + right_p < self.size[1]:
            w_pad = self.size[1] - w - left_p - right_p
            left_p += w_pad
            right_p += w_pad
        need_pad = (top_p + bot_p + left_p + right_p) > 0
        return top_p, bot_p, left_p, right_p, need_pad

    def _process_crop_op(self, imgs: List[matx.NDArray]) -> List[matx.NDArray]:
        batch_size = len(imgs)
        need_pad = False
        top_pads, bot_pads, left_pads, right_pads = [], [], [], []
        padded_height, padded_width = [], []
        for i in range(batch_size):
            h, w = imgs[i].shape()[:2]
            top_p, bot_p, left_p, right_p, need_pad_tmp = self.get_pad_params(h, w)
            need_pad |= need_pad_tmp
            top_pads.append(top_p)
            bot_pads.append(bot_p)
            left_pads.append(left_p)
            right_pads.append(right_p)
            padded_height.append(h + top_p + bot_p)
            padded_width.append(w + left_p + right_p)
        if need_pad:
            imgs = self.pad_op(imgs, top_pads, bot_pads, left_pads, right_pads, sync=self.sync)
        crop_x, crop_y, crop_w, crop_h = [], [], [], []
        for i in range(batch_size):
            y, x, h, w = self.get_crop_params(padded_height[i], padded_width[i])
            crop_x.append(x)
            crop_y.append(y)
            crop_w.append(w)
            crop_h.append(h)
        imgs = self.crop_op(imgs, crop_x, crop_y, crop_w, crop_h, sync=self.sync)
        return imgs

    def _process(self, imgs: List[matx.NDArray]) -> List[matx.NDArray]:
        return self._process_crop_op(imgs)

    def __repr__(self) -> str:
        return self.name + "(size={0}, padding={1}, device={2}, sync={3})".format(
            self.size, self.padding, self.device_str, self.sync)

File: vision/src/ops/cuda/crop.cc
// Copyright 2022 ByteDance Ltd. and/or its affiliates.
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

#include <cv_cuda.h>
#include <matxscript/runtime/native_object_registry.h>
#include <matxscript/runtime/py_args.h>
#include <mutex>
#include "matxscript/runtime/container/list_ref.h"
#include "matxscript/runtime/container/ndarray.h"
#include "matxscript/runtime/global_type_index.h"
#include "matxscript/runtime/logging.h"
#include "matxscript/runtime/runtime_value.h"
#include "utils/cuda/cuda_op_helper.h"
#include "utils/cuda/cuda_type_helper.h"
#include "vision_base_op_gpu.h"

namespace byted_matx_vision {
namespace ops {
namespace cuda {

using namespace matxscript::runtime;

class VisionCropOpGPU : public VisionBaseImageOpGPU<cuda_op::CropVarShape> {
 public:
  VisionCropOpGPU(const Any& session_info)
      : VisionBaseImageOpGPU<cuda_op::CropVarShape>(session_info) {
  }
  RTValue process(const List& arg_images,
                  const List& x,
                  const List& y,
                  const List& widths,
                  const List& heights,
                  int sync);
};

RTValue VisionCropOpGPU::process(const List& arg_images,
                                 const List& x,
                                 const List& y,
                                 const List& widths,
                                 const List& heights,
                                 int sync) {
  check_and_set_device(device_id_);
  auto images = check_copy(arg_images, ctx_, getStream());
  int batch_size = images.size();

  // calculate buffer
  size_t op_buffer_size = op_->calBufferSize(batch_size);
  std::shared_ptr<void> cpu_buffer_ptr(malloc(op_buffer_size), free);
  void* gpu_workspace = cuda_api_->Alloc(ctx_, op_buffer_size);

  // create event
  cudaStream_t cu_stream = getStream();
  cudaEvent_t finish_event;
  CHECK_CUDA_CALL(cudaEventCreate(&finish_event));

  auto finish_event_mutex = std::make_shared<std::mutex>();
  auto not_finish = std::make_shared<bool>(true);

  // construct outputs and inputs
  const void* input_ptr[batch_size];
  void* output_ptr[batch_size];
  cuda_op::DataShape input_shape[batch_size];
  cv::Rect roi[batch_size];
  int channel = 0;
  DataType nd_data_type;
  List res;

  int i = 0;
  for (const RTValue& elem : images) {
    auto view_elem = elem.AsObjectView<NDArray>();
    const NDArray& elem_data = view_elem.data();
    input_ptr[i] = (void*)(elem_data->data);
    std::vector<int64_t> src_shape = elem_data.Shape();

    if (i == 0) {
      channel = src_shape[2];
      nd_data_type = elem_data.DataType();
    } else {
      if (channel != src_shape[2]) {
        MXCHECK(false) << "Invalid input. The output shape should be equal";
      }
      if (nd_data_type != elem_data.DataType()) {
        MXCHECK(false) << "The inputs must have same data type";
      }
    }

    input_shape[i].N = 1;
    input_shape[i].C = channel;
    input_shape[i].H = src_shape[0];
    input_shape[i].W = src_shape[1];
    int tmp_x = x[i].As<int>();
    int tmp_y = y[i].As<int>();
    int tmp_width = widths[i].As<int>();
    int tmp_height = heights[i].As<int>();
    MXCHECK(0 <= tmp_x && 0 <= tmp_width && tmp_x + tmp_width <= src_shape[1])
        << "X + Width should be less than or equal to image width, but get : " << tmp_x + tmp_width
        << ", origin image width: " << src_shape[1];

    MXCHECK(0 <= tmp_y && 0 <= tmp_height && tmp_y + tmp_height <= src_shape[0])
        << "Y + Height should be less than or equal to image height, but get : "
        << tmp_y + tmp_height << ", origin image height: " << src_shape[0];

    roi[i].x = tmp_x;
    roi[i].y = tmp_y;
    roi[i].width = tmp_width;
    roi[i].height = tmp_height;
    std::vector<int64_t> out_shape = {tmp_height, tmp_width, channel};
    size_t output_buffer_size = CalculateOutputBufferSize(out_shape, nd_data_type);

    NDArray dst_arr = MakeNDArrayWithWorkSpace(
        ctx_,
        cuda_api_,
        output_buffer_size,
        out_shape,
        nd_data_type,
        [finish_event,
         finish_event_mutex,
         not_finish,
         elem_data,
         cpu_buffer_ptr,
         local_device_api = this->cuda_api_,
         local_device_id = this->device_id_,
         gpu_workspace]() {
          std::lock_guard<std::mutex> lock(*finish_event_mutex);
          if (*not_finish) {
            DLContext local_ctx;
            local_ctx.device_id = local_device_id;
            local_ctx.device_type = DLDeviceType::kDLCUDA;
            CHECK_CUDA_CALL(cudaEventSynchronize(finish_event));
            CHECK_CUDA_CALL(cudaEventDestroy(finish_event));
            local_device_api->Free(local_ctx, gpu_workspace);
            *not_finish = false;
          }
        },
        0,
        nullptr);
    res.push_back(dst_arr);
    output_ptr[i] = (void*)(dst_arr->data);
    i += 1;
  }
  // call kernel
  cuda_op::DataType op_data_type = DLDataTypeToOpencvCudaType(nd_data_type);
  op_->infer(input_ptr,
             output_ptr,
             gpu_workspace,
             (void*)cpu_buffer_ptr.get(),
             batch_size,
             op_buffer_size,
             roi,
             input_shape,
             cuda_op::kNHWC,
             op_data_type,
             cu_stream);
  // record stop event on the stream
  CHECK_CUDA_CALL(cudaEventRecord(finish_event, cu_stream));
  CUDA_EVENT_SYNC_IF_DEBUG(finish_event);
  CUDA_STREAM_SYNC_IF_DEBUG(cu_stream);
  CUDA_DEVICE_SYNC_IF_DEBUG();
  if (sync != VISION_SYNC_MODE::ASYNC) {
    CHECK_CUDA_CALL(cudaEventSynchronize(finish_event));
    if (sync == VISION_SYNC_MODE::SYNC_CPU) {
      return to_cpu(res, getStream());
    } else {
      return res;
    }
  }
  return res;
};

MATX_REGISTER_NATIVE_OBJECT(VisionCropOpGPU)
    .SetConstructor([](PyArgs args) -> std::shared_ptr<void> {
      MXCHECK_EQ(args.size(), 1) << "[BytedVisionCropOpGPU] Constructor Expect 1 arguments but get "
                                 << args.size();
      return std::make_shared<VisionCropOpGPU>(args[0]);
    })
    .RegisterFunction("process", [](void* self, PyArgs args) -> RTValue {
      MXCHECK_EQ(args.size(), 6) << "[BytedVisionCropOpGPU] Expect 6 arguments but get "
                                 << args.size();
      return reinterpret_cast<VisionCropOpGPU*>(self)->process(args[0].AsObjectView<List>().data(),
                                                               args[1].AsObjectView<List>().data(),
                                                               args[2].AsObjectView<List>().data(),
                                                               args[3].AsObjectView<List>().data(),
                                                               args[4].AsObjectView<List>().data(),
                                                               args[5].As<int>());
    });
}  // namespace cuda
}  // namespace ops
}  // namespace byted_matx_vision